tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1306 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+#include "absl/strings/escaping.h"
+#include "absl/strings/match.h"
+
+#include "tensorstore/internal/data_copy_concurrency_resource.h"
+#include "tensorstore/internal/http/curl_transport.h"
+#include "tensorstore/internal/http/http_request.h"
+#include "tensorstore/internal/http/http_response.h"
+#include "tensorstore/internal/http/http_transport.h"
+#include "tensorstore/internal/metrics/counter.h"
+#include "tensorstore/internal/metrics/histogram.h"
+#include "tensorstore/internal/retry.h"
+#include "tensorstore/internal/schedule_at.h"
+#include "tensorstore/internal/digest/md5.h"
+#include "tensorstore/internal/digest/sha256.h"
+#include "tensorstore/kvstore/byte_range.h"
+#include "tensorstore/kvstore/driver.h"
+#include "tensorstore/kvstore/generation.h"
+#include "tensorstore/kvstore/registry.h"
+#include "tensorstore/kvstore/spec.h"
+#include "tensorstore/kvstore/url_registry.h"
+#include "tensorstore/kvstore/gcs/validate.h"
+#include "tensorstore/kvstore/s3/s3_resource.h"
+#include "tensorstore/kvstore/s3/s3_request_builder.h"
+#include "tensorstore/kvstore/s3/s3_credential_provider.h"
+#include "tensorstore/kvstore/s3/s3_metadata.h"
+#include "tensorstore/kvstore/s3/validate.h"
+#include "tensorstore/kvstore/s3/s3_uri_utils.h"
+#include "tensorstore/util/execution/any_receiver.h"
+#include "tensorstore/util/execution/execution.h"
+#include "tensorstore/util/executor.h"
+#include "tensorstore/util/future.h"
+#include "tensorstore/util/garbage_collection/fwd.h"
+#include "tensorstore/util/quote_string.h"
+#include "tensorstore/util/result.h"
+#include "tensorstore/util/status.h"
+#include "tensorstore/util/str_cat.h"
+
+
+// specializations
+#include "tensorstore/internal/cache_key/std_optional.h"
+#include "tensorstore/internal/json_binding/std_array.h"
+#include "tensorstore/internal/json_binding/std_optional.h"
+#include "tensorstore/util/garbage_collection/std_optional.h"
+#include "tensorstore/serialization/fwd.h"
+#include "tensorstore/serialization/std_optional.h"
+
+
+using ::tensorstore::internal::IntrusivePtr;
+using ::tensorstore::internal::DataCopyConcurrencyResource;
+using ::tensorstore::internal::ScheduleAt;
+using ::tensorstore::internal::SHA256Digester;
+using ::tensorstore::internal::MD5Digester;
+using ::tensorstore::internal_http::HttpRequest;
+using ::tensorstore::internal_http::HttpResponse;
+using ::tensorstore::internal_http::HttpTransport;
+using ::tensorstore::internal_auth_s3::S3Credentials;
```

Comment:
we tend to use a namespace maps to path scheme, rather than proliferating namespaces, you could use internal_s3 or internal_kvstore_s3 or similar for all these, IMO.

Reply:
Used `internal_kvstore_s3`